optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/RearrangingBlockingQueue.java

```diff
@@ -0,0 +1,205 @@
+            if (filterStepIndex == stepIndex) {
+                throw new IllegalStateException("The filterStepIndex (" + filterStepIndex
+                        + ") cannot be the same as the stepIndex (" + stepIndex + ")");
+            }
+            filterStepIndex = stepIndex;
+            innerQueue.clear();
+        }
+        searchMoveIndex = 0;
+        backlog.clear();
+    }
+
+    /**
+     * This method is thread-safe. It can be called from any move thread.
+     * @param moveThreadIndex {@code 0 <= moveThreadIndex < moveThreadCount}
+     * @param stepIndex at least 0
+     * @param moveIndex at least 0
+     * @param move never null
+     * @see BlockingQueue#add(Object)
+     */
+    public void addUndoableMove(int moveThreadIndex, int stepIndex, int moveIndex, Move<Solution_> move) {
+        MoveResult<Solution_> result = new MoveResult<>(moveThreadIndex, stepIndex, moveIndex, move, false, null);
+        synchronized (this) {
+            if (result.getStepIndex() != filterStepIndex) {
+                // Discard element from previous step
+                return;
+            }
+            innerQueue.add(result);
+        }
+    }
+
+    /**
+     * This method is thread-safe. It can be called from any move thread.
+     * @param moveThreadIndex {@code 0 <= moveThreadIndex < moveThreadCount}
+     * @param stepIndex at least 0
+     * @param moveIndex at least 0
+     * @param move never null
+     * @param score never null
+     * @see BlockingQueue#add(Object)
+     */
+    public void addMove(int moveThreadIndex, int stepIndex, int moveIndex, Move<Solution_> move, Score score) {
+        MoveResult<Solution_> result = new MoveResult<>(moveThreadIndex, stepIndex, moveIndex, move, true, score);
+        synchronized (this) {
+            if (result.getStepIndex() != filterStepIndex) {
+                // Discard element from previous step
+                return;
+            }
+            innerQueue.add(result);
+        }
+    }
+
+    /**
+     * This method is thread-safe. It can be called from any move thread.
+     * Previous results (that haven't been consumed yet), will still be returned during iteration
+     * before the iteration throws an exception,
+     * unless there's a lower moveIndex that isn't in the queue yet.
+     * @param moveThreadIndex {@code 0 <= moveThreadIndex < moveThreadCount}
+     * @param throwable never null
+     */
+    public void addExceptionThrown(int moveThreadIndex, Throwable throwable) {
+        MoveResult<Solution_> result = new MoveResult<>(moveThreadIndex, throwable);
+        synchronized (this) {
+            innerQueue.add(result);
+        }
+    }
+
+    /**
+     * Not thread-safe. Can only be called from the solver thread.
+     * @return never null
+     * @throws InterruptedException if interrupted
+     * @see BlockingQueue#take()
+     */
+    public MoveResult<Solution_> take() throws InterruptedException {
+        searchMoveIndex++;
```

Comment:
@rsynek this causes a one-off issue that deadlocks moveThreadCount=1 with buffer=1, because the searchMoveIndex is initialized to 0, so the first one it's looking for is moveIndex 1, but the first moveIndex is 0.

Reply:
The question is, if this is what triggered PLANNER-1088...

Reply:
@ge0ffrey truth, this might be just the first step to solve it.